ethcore/src/miner/miner.rs

```diff
@@ -117,267 +110,185 @@ pub struct MinerOptions {
-			true => Vec::new(),
-			false => vec![Box::new(WorkPoster::new(&options.new_work_notify))],
-		};
+	/// Set a callback to be notified about imported transactions' hashes.
+	pub fn add_transactions_listener(&self, f: Box<Fn(&[H256]) + Send + Sync>) {
+		self.transaction_queue.add_listener(f);
+	}
 
-		let service_transaction_action = match options.refuse_service_transactions {
-			true => ServiceTransactionAction::Refuse,
-			false => ServiceTransactionAction::Check(ServiceTransactionChecker::default()),
-		};
+	/// Creates new instance of miner Arc.
+	pub fn new(options: MinerOptions, gas_pricer: GasPricer, spec: &Spec, accounts: Option<Arc<AccountProvider>>) -> Miner {
+		let limits = options.pool_limits.clone();
+		let verifier_options = options.pool_verification_options.clone();
+		let tx_queue_strategy = options.tx_queue_strategy;
 
 		Miner {
-			transaction_queue: Arc::new(RwLock::new(txq)),
-			transaction_listener: RwLock::new(vec![]),
-			next_allowed_reseal: Mutex::new(Instant::now()),
-			next_mandatory_reseal: RwLock::new(Instant::now() + options.reseal_max_period),
-			sealing_block_last_request: Mutex::new(0),
-			sealing_work: Mutex::new(SealingWork{
+			sealing: Mutex::new(SealingWork {
 				queue: UsingQueue::new(options.work_queue_size),
 				enabled: options.force_sealing
-					|| !options.new_work_notify.is_empty()
-					|| spec.engine.seals_internally().is_some()
+					|| spec.engine.seals_internally().is_some(),
+				next_allowed_reseal: Instant::now(),
+				next_mandatory_reseal: Instant::now() + options.reseal_max_period,
+				last_request: 0,
 			}),
-			gas_range_target: RwLock::new((U256::zero(), U256::zero())),
-			author: RwLock::new(Address::default()),
-			extra_data: RwLock::new(Vec::new()),
-			options: options,
-			accounts: accounts,
-			engine: spec.engine.clone(),
-			notifiers: RwLock::new(notifiers),
+			params: RwLock::new(AuthoringParams::default()),
+			listeners: RwLock::new(vec![]),
 			gas_pricer: Mutex::new(gas_pricer),
-			service_transaction_action: service_transaction_action,
+			options,
+			transaction_queue: Arc::new(TransactionQueue::new(limits, verifier_options, tx_queue_strategy)),
+			accounts,
+			engine: spec.engine.clone(),
 		}
 	}
 
-	/// Creates new instance of miner with accounts and with given spec.
-	pub fn with_spec_and_accounts(spec: &Spec, accounts: Option<Arc<AccountProvider>>) -> Miner {
-		Miner::new_raw(Default::default(), GasPricer::new_fixed(20_000_000_000u64.into()), spec, accounts)
-	}
-
-	/// Creates new instance of miner without accounts, but with given spec.
-	pub fn with_spec(spec: &Spec) -> Miner {
-		Miner::new_raw(Default::default(), GasPricer::new_fixed(20_000_000_000u64.into()), spec, None)
-	}
-
-	fn forced_sealing(&self) -> bool {
-		self.options.force_sealing || !self.notifiers.read().is_empty()
+	/// Creates new instance of miner with given spec and accounts.
+	///
+	/// NOTE This should be only used for tests.
```

Comment:
Why not mark it as `cfg(test)` then?

Reply:
could be extracted, but the reason is probably that not all relevant tests reside in the same crate and `cfg(test)` isn't replicated when compiling dependencies.